main.cc

```diff
@@ -966,14 +966,13 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
             supervisor::notify("creating and verifying directories");
             utils::directories::set dir_set;
-            dir_set.add(cfg->data_file_directories());
-            dir_set.add(cfg->commitlog_directory());
-            dir_set.add(cfg->schema_commitlog_directory());
-            dirs.emplace(cfg->developer_mode());
+            dir_set.add(dirs->get_data_file_dirs());
+            dir_set.add(dirs->get_commitlog_dir());
+            dir_set.add(dirs->get_schema_commitlog_dir());
             dirs->create_and_verify(std::move(dir_set)).get();
```

Comment:
This thing becomes buggy now. You have already captured all the directories from cfg in the constructor up above (the `dirs.emplace(cfg)` part). Now the dir_set duplicates the set, I think that the set should be replaced in favor of just using the specific dirs from directories themselves

Reply:
I had similar feeling about that. I was just not sure how to proceed. I preserved the logic.

1. Three field are used here to create directories via `utils::directories::set` + `create_and_verify()`.
2. Two fields are used below to create directories via `db::hints::directory_initializer` + `ensure_created_and_verified()`.

Therefore, I was not sure whether:
- `utils::directories` should create directories only for these three fields and the remaining two should be created via `directory_initializer` as it was
- `utils::directories` should create directories for all five fields and the usage of `directory_initializer` should be altered

I do not have a strong opinion, because I am not aware of many aspects related to the project. What do you think about it?

Reply:
I checked and it seems, that one of hints dir may be created conditionally.

In `main.cc` we check `hinted_handoff_enabled.is_disabled_for_all` and depending on that flag we create the directories.

In `api/storage_proxy.cc` I can see that the flag may be modified via request:
```
    sp::set_hinted_handoff_enabled.set(r, [&proxy](std::unique_ptr<http::request> req)  {
        auto enable = req->get_query_param("enable");
        auto filter = (enable == "true" || enable == "1")
                ? db::hints::host_filter(db::hints::host_filter::enabled_for_all_tag {})
                : db::hints::host_filter(db::hints::host_filter::disabled_for_all_tag {});
        return proxy.invoke_on_all([filter = std::move(filter)] (service::storage_proxy& sp) {
            return sp.change_hints_host_filter(filter);
        }).then([] {
            return make_ready_future<json::json_return_type>(json_void());
        });
    });
```

The handler for this request calls `sp.change_hints_host_filter()`. In case of storage proxy it may create the directories via dir initializer.

```
future<> storage_proxy::change_hints_host_filter(db::hints::host_filter new_filter) {
    if (new_filter == _hints_manager.get_host_filter()) {
        co_return;
    }

    co_await _hints_directory_initializer.ensure_created_and_verified();
    co_await _hints_directory_initializer.ensure_rebalanced();
    // This function is idempotent
    co_await _hints_resource_manager.register_manager(_hints_manager);
    co_await _hints_manager.change_host_filter(std::move(new_filter));
}
```

From my POV it seems, that the logic related to directories initializer needs to be preserved.

Reply:
I added a new function to `utils::directories` to encapsulate the creation of these three dirs.